### PR TITLE
Branch Key

### PR DESCRIPTION
**Pull request by qinweigong**
@aaustin 

- [ ] The branch_key in strings.xml is still temp. Need to change it once backend is deployed, and fix tests.
- [ ] Update README with sample code for setting both live and test branch key and switch between them.

### CODE REVIEW

**Comment by aaustin** [comment]
Do we want to add in support for test and live here, like allow the dev to enter both keys and flip a boolean somewhere? Were you thinking of adding this later

At the very least, let's put a warning when the session is initialized with a key that contains the test string.

Looks great otherwise!

File: Branch-SDK/src/io/branch/referral/Branch.java [hunk 1]
```diff
@@ -138,29 +138,40 @@ private Branch(Context context) {
...
-	public static Branch getInstance(Context context) {
+	private static Branch getBranchInstance(Context context, boolean isLive) {
 		if (branchReferral_ == null) {
 			branchReferral_ = Branch.initInstance(context);
 
-			String appKey = branchReferral_.prefHelper_.getAppKey();
-	        if (appKey == null || appKey.equalsIgnoreCase(PrefHelper.NO_STRING_VALUE)) {
-	        	Log.i("BranchSDK", "Branch Warning: Please enter your Branch App Key in your project's res/values/strings.xml!");
+			String branchKey = branchReferral_.prefHelper_.getBranchKey(isLive);
+	        if (branchKey == null || branchKey.equalsIgnoreCase(PrefHelper.NO_STRING_VALUE)) {
```
Comment (aaustin) [inline Branch-SDK/src/io/branch/referral/Branch.java]: let's make sure we don't throw warnings if app key is present

File: Branch-SDK/src/io/branch/referral/Branch.java [hunk 2]
```diff
@@ -478,7 +489,8 @@ public void setIdentity(final String userId) {
 			public void run() {
 				JSONObject post = new JSONObject();
 				try {
-					post.put("app_id", prefHelper_.getAppKey());
+					post.put(RemoteInterface.BRANCH_KEY, prefHelper_.getBranchKey());
+                    post.put("app_id", prefHelper_.getAppKey());
```
Comment (aaustin) [inline Branch-SDK/src/io/branch/referral/Branch.java]: feedback from dmitri about only including non blank one

File: Branch-SDK/src/io/branch/referral/Branch.java [hunk 3]
```diff
@@ -1460,6 +1484,14 @@ private void registerInstallOrOpen(String tag) {
 	}
 
 	private void initializeSession() {
+        if (prefHelper_.getBranchKey() == null || prefHelper_.getBranchKey().equalsIgnoreCase(PrefHelper.NO_STRING_VALUE)
+            && prefHelper_.getAppKey() == null || prefHelper_.getAppKey().equalsIgnoreCase(PrefHelper.NO_STRING_VALUE)) {
```
Comment (aaustin) [inline Branch-SDK/src/io/branch/referral/Branch.java]: think you need some more parenthesis around these () && ()

File: Branch-SDK/src/io/branch/referral/RemoteInterface.java [hunk 4]
```diff
@@ -119,10 +119,12 @@ public ServerResponse make_restful_post(JSONObject body, String url, String tag,
...
 	
 	public ServerResponse make_restful_post(JSONObject body, String url, String tag, int timeout, boolean log, BranchLinkData linkData) {
-		try {    	
-			if (body.has("app_id") && body.getString("app_id").equals(PrefHelper.NO_STRING_VALUE)) {
-				return new ServerResponse(tag, NO_API_KEY_STATUS);
-			}
+		try {
+			if (body.has(BRANCH_KEY) && body.getString(BRANCH_KEY).equals(PrefHelper.NO_STRING_VALUE)) {
+				return new ServerResponse(tag, NO_BRANCH_KEY_STATUS);
+			} else if (body.has("app_id") && body.getString("app_id").equals(PrefHelper.NO_STRING_VALUE)) {
```
Comment (aaustin) [inline Branch-SDK/src/io/branch/referral/RemoteInterface.java]: maybe combine this into () || ()

**Comment by aaustin** [comment]
Can you also update the docs with the getTestInstance and getInstance? maybe one commented out above the other with some explanation in the sample code?

File: Branch-SDK/src/io/branch/referral/Branch.java (hunk 2 above)
Comment (Sarkar) [inline Branch-SDK/src/io/branch/referral/Branch.java]: also, I don't know if it's github but this spacing is kind of funky

File: Branch-SDK/src/io/branch/referral/Branch.java (hunk 1 above)
Comment (qinweigong) [inline Branch-SDK/src/io/branch/referral/Branch.java]: Well, the purpose of this warning is prompt partners to transition from using app key to branch key, as we will abandon app key eventually. Will make sure the sdk is pushed out after backend.

File: Branch-SDK/src/io/branch/referral/Branch.java (hunk 3 above)
Comment (qinweigong) [inline Branch-SDK/src/io/branch/referral/Branch.java]: yikes! good catch!

File: Branch-SDK/src/io/branch/referral/RemoteInterface.java (hunk 4 above)
Comment (qinweigong) [inline Branch-SDK/src/io/branch/referral/RemoteInterface.java]: Yes. the check for app_id was added back for temporary backward compatibility and meant to be removed in the near future. But, temp code is still code :)

**Comment by qinweigong** [comment]
@aaustin Updated per your feedback and the README

**Comment by aaustin** [comment]
:+1: after the backend goes live

File: Branch-SDK/src/io/branch/referral/PrefHelper.java [hunk 5]
```diff
@@ -123,26 +125,56 @@ public String getAppVersion() {
...
+
+    public String getAppKey() {
+        String appKey = null;
+        try {
+            final ApplicationInfo ai = context_.getPackageManager().getApplicationInfo(context_.getPackageName(), PackageManager.GET_META_DATA);
+            if (ai.metaData != null) {
+                appKey = ai.metaData.getString("io.branch.sdk.ApplicationId");
+            }
+        } catch (final PackageManager.NameNotFoundException e) {
+        }
```
Comment (Sarkar) [inline Branch-SDK/src/io/branch/referral/PrefHelper.java]: Should we throw a Log.e here?

Comment (qinweigong) [inline Branch-SDK/src/io/branch/referral/PrefHelper.java]: doesn't really matter.

**Pull request merged**